### PR TITLE
[AIRFLOW-5930] Use cached-SQL query building for hot-path queries

### PR DESCRIPTION
**Pull request by ashb**
Make sure you have checked _all_ steps below.

### Jira

- [x] https://issues.apache.org/jira/browse/AIRFLOW-5930

### Description

- [x] Building the SQL string for this query takes up about 25% of the time that the DAG parsing process spends, so replacing this one query should help speed up the rate at which the scheduler can queue tasks.

  See https://docs.sqlalchemy.org/en/13/orm/extensions/baked.html for more info. The docs explain a lot of how/why this works, so rather than rebuilding them from string 10s of times (once per task per active dag run) we cache the build SQL string!

  I will collect up-to-date performance numbers against master, but this makes the "dag parsing" process of the scheduler (which creates and updates dag runs, and creates Task Instances) about 2x quicker:

  Concurrent DagRuns | Tasks | Before | After | Speedup
  -- | -- | -- | -- | --
  2 | 12 | 0.146s (±0.0163s) | 0.074s (±0.0037s) | x1.97
  10 | 12 | 1.11s (±0.0171s) | 0.266s (±0.0229s) | x4.17
  40 | 12 | 4.28s (±0.101s) | 0.852s (±0.0113s) | x5.02
  40 | 40 | 6.72s (±0.067s) | 2.659s (±0.0283s) | x2.53

### Tests

- [x] No new tests, no change in behaviour.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (mik-laj) [inline airflow/ti_deps/deps/trigger_rule_dep.py]: Can you provide me this query in SQL format? I think it can be optimized for PostgresQL by using COUNT...FILTER syntax. However, this also requires checking if this syntax has an effect on performance, or is it just syntactic sugar. But for logic this additional information can be used by the planner to make a more efficient query.
https://www.postgresql.org/docs/9.4/sql-expressions.html#SYNTAX-AGGREGATES

File: airflow/__init__.py
Comment (mik-laj) [inline airflow/__init__.py]: I don't know if this should be done here or when starting SchedulerJob.  In my opinion, adding additional logic to init is not the best solution and we can probably avoid it in this situation. We don't need this query to be loaded in many cases, e.g. on workers.

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [inline airflow/jobs/scheduler_job.py]: Can you also add rtype for _process_task_instances also? Now it is difficult to check if this is true. Especially since this method previously use TaskInstanceKeyType.

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [inline airflow/jobs/scheduler_job.py:1000]: Why did you delete this decorator? It has no effect on performance because it is very simple logic.

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [inline airflow/jobs/scheduler_job.py:1058]: I really don't like filtering with the like expression. This makes the query very difficult to optimize. It is not possible to store it in a simple data structure. We have to have a very complex binary tree, but which takes more memory than a simple structure with 3 values. Which causes other problems, e.g. unbalanced tree, and thus performance degradation.

File: airflow/models/dagrun.py
Comment (mik-laj) [inline airflow/models/dagrun.py:296]: Is it not necessary to change the method name? Now does not contain information about tasks. This may not be clear in the future.

File: airflow/models/dagrun.py
Comment (mik-laj) [inline airflow/models/dagrun.py]: Do you think it is worth dividing the loop from line 272 into two loops? One loop will filters the elements and the second loop will set tasks on task instances. This does not affect performance, but will make it easier to understand the code.

File: airflow/models/dagrun.py
Comment (mik-laj) [inline airflow/models/dagrun.py]: Can you check if double calling get_task_instances is faster than filtering the list in Python? Line 319 and 306 contains calls to the get_task_instances method, and this method invokes a database query.

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [inline airflow/jobs/scheduler_job.py:1000]: This method has an invalid rtype. Returns two dictionaries in a tuple, not just one dictionary. Can you correct that?

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [inline airflow/jobs/scheduler_job.py]:
```suggestion
```
This method doesn't have to return anything. The user of this method can call len () on one parameter of this method. This will always give the same result. 

> Explicit is better than implicit.

Python Zen

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [inline airflow/jobs/scheduler_job.py]: Why do we need enumerate if the ``i`` variable  is not used inside the loop?

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [inline airflow/jobs/scheduler_job.py]: Doesn't this method return TaskInstance now?  We should pass TaskInstance instead of TaskInstance key from what I was checking recently. I wanted to improve it in the near future.  Earlier, we returned the key because we had multiprocessing.

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [inline airflow/jobs/scheduler_job.py:1344]: This variable name is not valid because this method now returns TaskInstance rather than SimpleTaskInstance.

Comment (mik-laj) [reply airflow/jobs/scheduler_job.py:1344]: https://github.com/apache/airflow/pull/6792/files#diff-c35269bcfbbe386e269ffa7487e86192R1238

File: airflow/__init__.py
Comment (mik-laj) [reply airflow/__init__.py]: What do you think about? 
https://github.com/apache/airflow/blob/59fbc2f84a96c1677c8e420d7ce761f1d82b06ac/airflow/jobs/scheduler_job.py#L147

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py:1000]: Because I _want_ to force a session to be reused.

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py:1058]: I agree, and we should store a "run_type" column or something similar on the dag run table.

But this is unchanged behaviour.

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py]: I'll double check, but I think this avoided a `count()` query when we already new the answer!

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py]: Yes, again this was deliberate.

We went from TI, to TIKey, and then back to TI when it was used! This stops an extra DB lookup.

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (ashb) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: Yeah, I was wondering about having some DB specific optimizations in places. Somewhere else in the Scheduler we can speed it up by doing `UPDATE task_instance ... RETURNING *` to avoid a second query, I think it helped a bit.

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [reply airflow/jobs/scheduler_job.py]: In the previous version, we use len() on list. It doesn't look like it needs optimisation.
https://github.com/apache/airflow/pull/6792/files/59fbc2f84a96c1677c8e420d7ce761f1d82b06ac#diff-c35269bcfbbe386e269ffa7487e86192L1355

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [reply airflow/jobs/scheduler_job.py]: Do we want to work on TaskInstance objects all the time?  _process_dags method return List of TaskInstance, but we assign the result to a variable of type ti_key.  This is confusing.

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py]: Hmmm yes. Maybe this changed in the rebasing, or I'm making stuff up :grin:

File: airflow/__init__.py
Comment (ashb) [reply airflow/__init__.py]: (thought I already left this comment, but can't see it)

Yes, this is a hack and needs fixing before this is ready to merge. Putting it somehere in SchedulerJob (or DagFileProcessorAgent perhaps) might be a good way, as that is the only time we need to prime the cache for this specific query.

File: airflow/models/dagrun.py
Comment (ashb) [reply airflow/models/dagrun.py]: Will check. I was mostly making these changes based on the hot-spots identified by the profling, but I have already extended this a bit beyond just that with the refactor of this method.

Perhaps I should revert some of these changes and make them in to a separate PR?

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py]: Used by the `return` -- basically this avoids iterating over the loop, and then counting the length of the loop again.

_This_ might have been why I wanted to avoid the `len()` call, in my large dags there are 1000s of dags. But I suspect this was me just making stuff up that isn't really needed :D

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py]: Oh, this was a mistake in my rebase. This should be `for ti in self._process_dags`

File: airflow/models/dagrun.py
Comment (mik-laj) [reply airflow/models/dagrun.py]: Scheduler's performance is also an area of interest for Polidea and our clients, so dividing work into smaller PRs will allow us to develop better solutions. together

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [reply airflow/jobs/scheduler_job.py:1058]: I created ticket: 
https://issues.apache.org/jira/browse/AIRFLOW-6242
A small step forward.

File: airflow/models/dagrun.py
Comment (ashb) [reply airflow/models/dagrun.py:296]: I don't think so - nothing about it changed, it looks at the TIs associated with the DagRun (based on `self.get_task_instances()`)

File: airflow/models/dagrun.py
Comment (ashb) [inline airflow/models/dagrun.py]: This function showed up as a noticable chunk of the profling time for the DagFileProcessor, so we should decide if it's needed.

Right now as it is (and was) called it re-used the same session, so would be run in the same transaction. I think this makes the call pointless as it's asking for the state of the TI, but since its in the same transaction it will have a consistent view of the row, right?

I'm basing this off https://www.postgresql.org/docs/9.5/transaction-iso.html which says 

> Read Committed is the default isolation level in PostgreSQL. When a transaction uses this isolation level, a SELECT query (without a FOR UPDATE/SHARE clause) sees only data committed before the query began; it never sees either uncommitted data or changes committed during query execution by concurrent transactions. In effect, a SELECT query sees a snapshot of the database as of the instant the query begins to run

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py]: Yeah no, this is not needed. Reverthing this change.

**Comment by shlomiken** [comment]
@ashb - you probably on vacation now , but i wanted to see if you can help so i would be able to make progress with this image you have provided us . astronomerio / ap-airflow : 1.10.5-perf-fixes-alpine3.10-onbuild
the problem is that i first got some module not found errors and now i get those messages .
` File "/usr/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 581, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) relation "connection" does not exist
LINE 2: FROM connection GROUP BY connection.conn_id
             ^
`
i really want to test this as soon as possible , but cannot tell what is wrong with this image.
thanks
Shlomi

**Comment by shlomiken** [comment]
> @ashb - you probably on vacation now , but i wanted to see if you can help so i would be able to make progress with this image you have provided us . astronomerio / ap-airflow : 1.10.5-perf-fixes-alpine3.10-onbuild
> the problem is that i first got some module not found errors and now i get those messages .
> `File "/usr/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 581, in do_execute cursor.execute(statement, parameters) sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) relation "connection" does not exist LINE 2: FROM connection GROUP BY connection.conn_id ^`
> i really want to test this as soon as possible , but cannot tell what is wrong with this image.
> thanks
> Shlomi

I think that the connection issue is behind me - was related to astro-cli version i guess.
but now i see this one 
`Traceback (most recent call last):
  File "/usr/bin/airflow", line 32, in <module>
    args.func(args)
  File "/usr/lib/python3.7/site-packages/airflow/utils/cli.py", line 74, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib/python3.7/site-packages/airflow/bin/cli.py", line 1017, in scheduler
    job.run()
  File "/usr/lib/python3.7/site-packages/airflow/jobs/base_job.py", line 213, in run
    self._execute()
  File "/usr/lib/python3.7/site-packages/airflow/jobs/scheduler_job.py", line 1317, in _execute
    from airflow.ti_deps.deps.trigger import TriggerRuleDep
ModuleNotFoundError: No module named 'airflow.ti_deps.deps.trigger'`

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (ashb) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: May not be needed after #4751 anyway.

**Comment by ashb** [comment]
Hi @shlomiken Sorry, my inbox is a disaster area right now :) I've fixed that issue, and a few more.

Something is still up with the tests here on master that I need to dig in to, but the backported version of this change to 1.10.5 should be ready tomorrow/Monday: https://travis-ci.com/astronomer/airflow/builds/143977776

**Comment by ashb** [comment]
Tests should be passing now, and we've been running a back-ported version of this for a few weeks and it makes a difference for us.

File: airflow/cli/commands/dag_command.py
Comment (madison-ookla) [inline airflow/cli/commands/dag_command.py]: I think the pylint disable can be removed here as a result of this change

Comment (ashb) [reply airflow/cli/commands/dag_command.py]: Oh yes good spot!

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [inline airflow/jobs/scheduler_job.py:1275]:
```suggestion
    def _enqueue_task_instances_with_queued_state(self, simple_dag_bag, simple_task_instances):
```
In my opinion, SimpleTaskInstance is processed here. You also forgot to update the method documentation after changing the name of the argument in the code. ;-)

File: airflow/jobs/scheduler_job.py
Comment (mik-laj) [inline airflow/jobs/scheduler_job.py:1312]:
```suggestion
            self.executor.queue_task_instance(
```
Can we also change it?

File: airflow/models/dagrun.py
Comment (mik-laj) [inline airflow/models/dagrun.py]:
```suggestion
        :rtype: Tuple[State, List[airflow.models.TaskInstance])
```

**Review by mik-laj — APPROVED** [review]

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py:1275]: This is called with the result of `self._change_state_for_executable_task_instances` which used to return this

```
        simple_task_instances = [SimpleTaskInstance(ti) for ti in
                                 tis_to_set_to_queued]
```

But one of the slow downs I noticed was continually re-looking up the TI, so it has been changed to pass full TaskInstance objects around (as we are all in the same process).

I'll update the docs to reflect the new code/type

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py:1312]: I guess we should

**Comment by ashb** [comment]
Tests aren't quite passing -- one failure due to a non-fixed ordering of return values.

**Review by turbaszek — APPROVED** [review]

File: airflow/models/dagrun.py
Comment (ashb) [reply airflow/models/dagrun.py]: For sphinx docs it should be `tuple` and `list`.

Comment (ashb) [reply airflow/models/dagrun.py]: This is what the current version renders as:

image

Comment (mik-laj) [reply airflow/models/dagrun.py]: It seemed to me that valid PEP types were supported. In that case, it probably should be `tuple[State, list[airflow.models.TaskInstance]]`

Comment (ashb) [reply airflow/models/dagrun.py]: Yup, that's what I'm changing it too. Link for reference http://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists (at the end of that section, just before "Cross-referencing Python objects")

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py:1312]: Oh queue_task_instance already exists -- I forgot you(?) added that already

Comment (ashb) [reply airflow/jobs/scheduler_job.py:1312]: This is actually more complex than it seems -- calling queue_task_instance sends it via `ti.command_as_list`, but to get the same behaviour with filepath we'd need to have ti.task.dag be set, but the ti doesn't have a task set.

Comment (ashb) [reply airflow/jobs/scheduler_job.py:1312]: Fixing this properly will also fix https://issues.apache.org/jira/browse/AIRFLOW-3877

Comment (ashb) [reply airflow/jobs/scheduler_job.py:1312]: I've put a TODO comment in here. Not happy about doing that but I think this PR is big enough already.

Comment (mik-laj) [reply airflow/jobs/scheduler_job.py:1312]: > Oh queue_task_instance already exists -- I forgot you(?) added that already

I wanted to add the queue_simple_task_instance method, but my PR is WIP.

**Comment by ashb** [comment]
Updated numbers from master branch. I'm not sure these will equate to _such_ a speed up in terms of actual task execution. Overall numbers are lower as I have a faster machine to run tests on now :)


  Concurrent DagRuns | Tasks | Before | After | Speedup
  -- | -- | -- | -- | --  
 100 | 12 | 3.9545s (±0.140s) | 1.7174s (±0.045s) | x2.3
 100 | 40 | 12.1224s (±0.357s) |9.0649s (±0.603s) | x1.3
 1000 | 12 | 17.7790s (±0.240s) | 8.3937s (±0.516s) | x1.5
 1000 | 40 | 67.1121s (±0.799s) | 44.6126s (±0.518s) | x1.5

**Comment by ashb** [comment]
This PR may not be needed so much anymore once PRs such as #7477 land (this would still make _a bit_ of difference, but possibly not as much anymore.)

I will re-investigate this in a week or two.

**Pull request closed without merging**
